jauntin/cybersource
# Remove cybersource package, use own requests

### What does this PR address?

Cybersource requires updating http signatures as noted in https://github.com/jauntin/cybersource/issues/2.

This PR also addresses the problem of relying on the cybersource PHP rest client package, which has been difficult to use, with package incompatibilities, breaking changes, etc. We will no longer use that package.




### How does this PR address the issue?

- Updates cybersource http signatures to avoid using parentheses
- Replaces the cybersource package with custom requests. Since jauntin only uses 3 request types, we are only replacing those specific requests
- Note: there are no breaking changes in this update

Other changes:

- Adds laravel pint in place of phpcs (using PSR12 config)
- Updates tests, adds 100% coverage
- Changes how we simulate declines in the sandbox environment, so it relies on environment variables




### Demo

This package logs requests:

```
local.INFO: cybersource.pay GBLI0000003: Begin request .  
local.INFO: cybersource.pay GBLI0000003: Request successful {"status":201} .  


local.INFO: cybersource.refund GBLI0000005, 7005830917466472204951: Begin request .
local.INFO: cybersource.refund GBLI0000005, 7005830917466472204951: Request failed {"request":{"paymentRequestId":"7005830917466472204951","referenceNumber":"GBLI0000005","currency":"USD","totalAmount":"125.00"},"response":{"status":400,"body":"{\"id\":\"7005831273556497804951\",\"submitTimeUtc\":\"2023-11-21T16:12:07Z\",\"status\":\"INVALID_REQUEST\",\"reason\":\"INVALID_DATA\",\"message\":\"Declined - One or more fields in the request contains invalid data\",\"details\":[{\"field\":\"paymentInformation.card.expirationMonth\",\"reason\":\"INVALID_DATA\"}]}"}} . 
```

Testing environments can trigger invalid requests to cybersource:

<img width="1253" alt="image" src="https://github.com/jauntin/cybersource/assets/3991125/b4750d7f-0623-4ed2-a12e-9f2d3c3add95">




### Manual testing (:tophat:) instructions

For developers testing before initial deployment: set up by checking out this branch and then linking this package to an app. For example, if you are running composer link: `composer link /workspace/cybersource`; or if not manually symlink this package into your vendor folder.

To test. In all apps using this package,

1. Test payments and refunds. Confirm that they work as expected in the user UI, and that logging is as expected.
2. Test failed payments and refunds where possible.




### Pre-merge Checklist

- [x] PR has been linked to an issue
- [x] Tests created or updated as necessary
- [x] Author has manually tested the update
- [ ] Reviewer has manually tested the update
- [x] Upon merge, commit history is clean
